### PR TITLE
Hof arch addendums

### LINKED ISSUE #19: HOF recommendations
For functions that can be written with iter OR map OR fold_right, recommend that they be done that way.

### PR DESCRIPTION
**Pull request by Vighnesh-V**
Lints Hofs. Resolves #19

### CODE REVIEW

File: lib/canonical/.#canonical.ml [hunk 1]
```diff
@@ -0,0 +1 @@
+vighnesh@vagelos-ve703-13685.apn.wlan.private.upenn.edu.42042
```
Comment (KeenWill) [inline lib/canonical/.#canonical.ml:1]: what is this file for?

File: lib/style/.#check.mli [hunk 2]
```diff
@@ -0,0 +1 @@
+vighnesh@vagelos-ve703-13685.apn.wlan.private.upenn.edu.42042
```
Comment (KeenWill) [inline lib/style/.#check.mli:1]: Same thing here

File: lib/style/match.ml [hunk 3]
```diff
@@ -1,36 +1,39 @@
 open Canonical
 open Utils
 open Astutils
-
+    open Check
```
Comment (KeenWill) [inline lib/style/match.ml:4]: indentation nit
```suggestion
open Check

```

**Comment by KeenWill** [comment]
It also appears that there are no HOF tests that fail; I assume this is because we haven't written all of the HOF tests yet right?

File: lib/canonical/.#canonical.ml (hunk 1 above)
Comment (Vighnesh-V) [reply lib/canonical/.#canonical.ml:1]: removed these files lol

File: lib/style/.#check.mli (hunk 2 above)
Comment (Vighnesh-V) [reply lib/style/.#check.mli:1]: this one too

**Comment by Vighnesh-V** [comment]
Changes made! This PR just tweaks our architecture a little, to enable us to write different kinds of linter rules over different chunks of the(any) AST. This just lays the foundation to start writing the hof recommendations - I'll do that in another PR.

**Comment by Vighnesh-V** [comment]
We now also have linting rules for map and iter.
Fold is a little more fucky at the moment, probably requiring work with the typed tree.

**Comment by Vighnesh-V** [comment]
Resolved the issue of distinguishing fold. We now have three rules - one for fold, one for map, one for iter that all work well on obvious cases (definitely more testing needed).
Screen Shot 2020-05-22 at 6 34 00 PM

Resolves #19

**Pull request merged**